// @firebase/firestore/dist/memory/rn/index.js
import firebase from '@firebase/app';
import '@firebase/util';
import '@firebase/logger';
import '@firebase/webchannel-wrapper';
import { K as Ku, e as ea, j, B as Bu, G as Gu, z as zu, J as Ju, t as th, n as nh, a as eh, i as ih, r as rh, c as ch, b as ah, d as ju, Z as Za, f as tu, U as Uu } from './prebuilt.rn-5f435a7b.js';
import { Component } from '@firebase/component';

const name = "@firebase/firestore";
const version = "2.2.0";

/**
 * @license
 * Copyright 2017 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const firestoreNamespace = {
    Firestore: Ku,
    GeoPoint: ea,
    Timestamp: j,
    Blob: Bu,
    Transaction: Gu,
    WriteBatch: zu,
    DocumentReference: Ju,
    DocumentSnapshot: th,
    Query: nh,
    QueryDocumentSnapshot: eh,
    QuerySnapshot: ih,
    CollectionReference: rh,
    FieldPath: ch,
    FieldValue: ah,
    setLogLevel: ju,
    CACHE_SIZE_UNLIMITED: Za
};
/**
 * Configures Firestore as part of the Firebase SDK by calling registerService.
 *
 * @param firebase - The FirebaseNamespace to register Firestore with
 * @param firestoreFactory - A factory function that returns a new Firestore
 *    instance.
 */
function configureForFirebase(firebase, firestoreFactory) {
    firebase.INTERNAL.registerComponent(new Component('firestore', container => {
        const app = container.getProvider('app').getImmediate();
        return firestoreFactory(app, container.getProvider('auth-internal'));
    }, "PUBLIC" /* PUBLIC */).setServiceProps(Object.assign({}, firestoreNamespace)));
}

/**
 * @license
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Registers the memory-only Firestore build with the components framework.
 */
function registerFirestore(instance) {
    configureForFirebase(instance, (app, auth) => new Ku(app, new tu(app, auth), new Uu()));
    instance.registerVersion(name, version);
}
registerFirestore(firebase);

export { registerFirestore };
